add 'Save' and 'Save As' options

pionere opened the pull request:
'fixes' #19

AJenbo commented:
Nice :) By the way besides fixes there are other trigger words like "resolves" or "closes" if you find that more fitting.

While testing I found couple of issues:
- Save format appears to be related to the opened file name and not the saved name (opening priest8.cel and saving it as test.cl2 saves cel data to the target file, opening golema.cl2 and saving it as test.cel saves cl2 data to the target file).
- Changing clipping when saving appears to produce empty or invalid files (tested when opening golema.cl2)
- Changing group count updates the group count, but does not redistribute the frames like in the saved file (maybe it would be better to handle this when importing, or allow editing in the viewer rather then have it as a save option?).
- It is possible to save to a file of any name, preferably we should enforce the extension to fit the data format (add default if none recognized (monster -> monster.cl2)

pionere commented:
> Nice :) By the way besides "fixes" there are other trigger words like "resolves" or "closes" if you find that more fitting.
> 
> While testing I found couple of issues:
> 
> * Save format appears to be related to the opened file name and not the saved name (opening priest8.cel and saving it as test.cl2 saves cel data to the target file, opening golema.cl2 and saving it as test.cel saves cl2 data to the target file).

Changing format is not supported. The filter is set appropriately and you bypassed it deliberately. See comments below for 'save to a file of any name'.

> * Changing clipping when saving appears to produce empty or invalid files (tested when opening golema.cl2)

It is not empty, but if you change the clipping the application can no longer open it, because it expects clipped cl2 files. You can open it by Open As with set clipping mode and width.

> * Changing group count updates the group count, but does not redistribute the frames like in the saved file (maybe it would be better to handle this when importing, or allow editing in the viewer rather then have it as a save option?).

It redistributes the frames. The issue was that D1Cl2 did not clear the groupFrameIndices on load. Not sure how this worked till now, but fixed it in line 226 of [d1cl2.cpp](https://github.com/savagesteel/d1-graphics-tool/pull/78/files#diff-3f60d102c861d850747d07718c1686528c85cd8ef0202cf40cca3e43c7a31bfc).

> * It is possible to save to a file of any name, preferably we should enforce the extension to fit the data format (add default if none recognized (monster -> monster.cl2)

Only if you deliberately bypass the filter, which is not possible on linux. So I'm not really sure how many checks do you expect to be built into applications like these... Added an addition check to prevent it for the moment ...

> * amp-filse are about 3.5x the size they should be. It looks like they are generated based on sub-tiles rather then tiles

well, maybe they should not allocate that many entries... Changed line 349 of [mainwindow.cpp](https://github.com/savagesteel/d1-graphics-tool/pull/78/files#diff-211f3f5927f01c141c069e947bb2c55143370f7269f7f1be024d0b21b01c4bc6).

AJenbo commented:
> Only if you deliberately bypass the filter, which is not possible on linux. So I'm not really sure how many checks do you expect to be built into applications like these... Added an addition check to prevent it for the moment ...

I'm on linux and it just lets me enter any name i didn't have to bypass anything.

pionere commented:
> I'm on linux and it just lets me enter any name i didn't have to bypass anything.

How? Was the filter not set? The edit field is read-only. With the browse button the filter should limit the names. Last time I tried to enter something which did not match the filter, it simply added an additional extension to it.

AJenbo commented:
![image](https://user-images.githubusercontent.com/204594/209870435-e22e36c5-85b3-458b-ab39-e7cbfa95d55e.png)

pionere commented:
well, that kinda stupid. I thought at linux was smart enough...

AJenbo commented:
There are several different desktop environments which probably behaves different. This is Gnome 3 I think

AJenbo commented on source/d1celbase.h:
```diff
@@ -4,9 +4,17 @@
 #include "openasdialog.h"
+#include "saveasdialog.h"
+
+#define SUB_HEADER_SIZE 0x0A
+#define CEL_BLOCK_HEIGHT 32
```
Nice with the named values :+1:  Should we maybe use `constexpr` rather then `#define`?

pionere replied:
I used `#define`, because that is used everywhere else.

AJenbo commented on source/exportdialog.cpp:
```diff
@@ -18,9 +18,15 @@ ExportDialog::~ExportDialog()
     this->cel = c;
+    this->min = m;
+    this->til = t;
+    this->sol = s;
+    this->amp = a;
```
sol and amp are now unused in this file and can be removed

pionere replied:
It seems `exportLevelSubtiles` checks if `sol` is valid. Since it is not really performance critical I think it is better to leave it for the moment, because the code is very much in the expanding phase.

AJenbo commented on source/saveasdialog.h:
```diff
@@ -0,0 +1,55 @@
+class D1CelBase;
+
+enum class SAVE_CLIPPING_TYPE {
+    CLIPPED_AUTODETECT,
```
Should this be `CLIPPED_PRESERVED`?

pionere replied:
I do not know yet. Maybe it can be merged with `OPEN_CLIPPING_TYPE` later. Maybe one of them needs to be extended later... I think it should be decided after some feedback from the users.

pionere commented:
> There are several different desktop environments which probably behaves different. This is Gnome 3 I think

Yeah, I still think it is a very bad idea. I set the name filter for a reason. This behavior could become a security issue quite easily...

AJenbo commented:
I don't see this being a security issue, you should never trust user input any way. In any case there isn't really anything we can do to change how it currently works even if we think it's bad, it's just how it works on some platforms

pionere commented:
> I don't see this being a security issue, you should never trust user input any way.

Right and there should not writes over memory arrays. And I'm sure I'm the first one who assumed that setting a name-filter actually filters the input and not just a hint to the user...

AJenbo commented on source/d1amp.cpp:
```diff
@@ -57,6 +53,37 @@ bool D1Amp::load(QString filePath, int allocate)
 
+bool D1Amp::save(SaveAsParam *params)
+{
+    QString filePath = this->getFilePath();
+    if (params != nullptr && !params->ampFilePath.isEmpty()) {
+        filePath = params->ampFilePath;
+        if (QFile::exists(filePath)) {
+            QMessageBox::StandardButton reply;
+            reply = QMessageBox::question(nullptr, "Confirmation", "Are you sure you want to overwrite the AMP file?", QMessageBox::Yes | QMessageBox::No);
+            if (reply != QMessageBox::Yes) {
+                return false;
+            }
+        }
+    }
+
+    QFile outFile = QFile(filePath);
+    if (!outFile.open(QIODevice::WriteOnly | QFile::Truncate)) {
+        QMessageBox::critical(nullptr, "Error", "Failed open file: " + filePath);
+        return false;
+    }
```
Looks like there are 7 almost identical copies of this. Could we maybe extract it in to a common function if we change `"Are you sure you want to overwrite the AMP file?"` to something like `"Are you sure you want to overwrite {}?", filePath´?

pionere replied:
I don't see much point in it. It does not really cover a real functionality, it just bunch of lines which are common. Besides `filePath`, `params` and `params->ampFilePath` should be passed to that common function. At that point the whole function is gutted.

AJenbo replied:
To avoid duplication, make the save functions more concise and specific to the file data.
It looks like the function could return `std::optional<QFile>`, in case of an empty response the save would do an early return.

Something roughly along this line:
```cpp
std::optional<QFile> getValidSaveOutput(QString filePath, QString selectedPath) {
    if (!selectedPath.isEmpty()) {
        filePath = selectedPath;
        if (QFile::exists(filePath)) {
            QMessageBox::StandardButton reply;
            reply = QMessageBox::question(nullptr, "Confirmation", "Are you sure you want to overwrite " + filePath + "?", QMessageBox::Yes | QMessageBox::No);
            if (reply != QMessageBox::Yes) {
                return {};
            }
        }
    }

    QFile outFile = QFile(filePath);
    if (!outFile.open(QIODevice::WriteOnly | QFile::Truncate)) {
        QMessageBox::critical(nullptr, "Error", "Failed open file: " + filePath);
        return {};
    }
    
    return outFile;
}
```

And then the call becomes:
```cpp
    QString selectedPath = params != nullptr ? params->ampFilePath : "";
    std::optional<QFile> outFile = getValidSaveOutput(this->getFilePath(), selectedPath);
    if (!outFile) {
        return false;
    }
```

AJenbo replied:
The current build error can probably be fixed by `QDataStream out(&outFile);` -> `QDataStream out(&outFile->get());` (haven't tested it, so  mostly a guess)

AJenbo commented on source/d1min.cpp:
```diff
@@ -58,6 +59,42 @@ bool D1Min::load(QString filePath, quint16 subtileCount)
+        }
+    }
+
+    this->minFilePath = filePath; // this->load(filePath, subtileCount);
```
I'm a little confused by this comment, is it just to say that we are updating the value originally set in `load()`?

pionere replied:
at the moment overwriting the minFilePath is enough, but in the future a whole (re-)load might be necessary (similar to d1cel.cpp)

AJenbo commented on source/d1amp.cpp:
```diff
@@ -2,11 +2,7 @@
-{
-    this->load(path);
-}
+#include <QMessageBox>
```
I don't think this import is needed now that the dialogs have been moved to `getValidSaveOutput()`

AJenbo commented:
In case you missed this reply:
> The current build error can probably be fixed by QDataStream out(&outFile); -> QDataStream out(&outFile->get()); (haven't tested it, so mostly a guess)

AJenbo commented on source/d1cl2.cpp:
```diff
@@ -289,6 +291,213 @@ bool D1Cl2::load(QString filePath, OpenAsParam *params)
+    QString filePath = this->getFilePath();
+    if (params != nullptr && !params->celFilePath.isEmpty()) {
+        filePath = params->celFilePath;
+        /*if (QFile::exists(filePath)) {
```
I might have missed something, but why do we not want to check if the file is being overwritten here?

pionere replied:
Because QFileDialog already does that.

AJenbo commented on source/d1amp.cpp:
```diff
@@ -57,6 +53,25 @@ bool D1Amp::load(QString filePath, int allocate)
+    }
+
+    // write to file
+    QDataStream out(*outFile);
```
You probably need to first grab the content of the optional and then get a pointer for it:
```suggestion
    QDataStream out(&*outFile);
```
or
```suggestion
    QDataStream out(&outFile->get());
```

pionere replied:
How do you want to retrieve the selected file-path?

AJenbo replied:
Finished revewing things, I'll take a look at this now :)

AJenbo replied:
I pushed a fix to the branch, hope you are ok with doing it that way.

AJenbo commented on source/d1cel.cpp:
```diff
@@ -138,3 +140,231 @@ bool D1Cel::load(QString filePath, OpenAsParam *params)
+    }
+
+    bool result;
+    // if ((params == nullptr && this->getGroupCount() > 1) || (params != nullptr && params->groupNum != 0)) {
```
Is this for a future case where we want to be able the edit the groups in the viewer UI rather then provide it as a paramter in the save dialog?

pionere replied:
That was the original test, but switched to the `this->type` solution. I'm still not exactly sure how to solve it exactly. Like do we need to handle a case where the group count is 1, but the user still wants to store it with a group-header?

AJenbo replied:
I went over all the D1 and HF .cel files and all of the compilation files have multiple groups (8) so I don't think there is a need to handle such a case.

Really the only reason to save to the .cel format is for levels and to overwrite existing assets without altering code. In all other cases .cl2 would be a better choice.

pionere replied:
I think we should add an option to create new cel/cl2 files. And if we do, we have to handle any kind of half-ready assets...

AJenbo replied:
Yes there should defiantly be a way to create a new tileset/sprite. But that doesn't mean we can't match the de facto standard of only using the composite format if there are multiple groups.

AJenbo commented on source/d1celtilesetframe.cpp:
```diff
@@ -158,3 +160,292 @@ void D1CelTilesetFrame::LoadRightTrapezoid(QByteArray &rawData)
+        for (x = MICRO_WIDTH - i; x < MICRO_WIDTH; x++) {
+            pixel = this->getPixel(x, y);
+            if (!pixel.isTransparent()) {
+                QMessageBox::critical(nullptr, "Error", "Invalid non-transparent pixel in a Right Triangle frame I.");
```
For now this is ok, but we should probably think about either forcing it or having a validation step so that the user won't be presented with potentially hundreds of dialogs and instead simply abort if invalid data is found before we start saving.

Come to think of it we could probably prevent such a case from even happening by determining the frame type based on the loaded data and not allowing the user to directly manipulate it.

pionere replied:
Well, this is an error-case. If this happens then something is really wrong.

AJenbo replied:
Maybe an assert would be the proper thing then?

pionere replied:
I would prefer to have a 'log' during save/export and dump such messages there. Like an extended progress bar with 'Details' button and a collapsible box which could be checked during and at the end of the process.

But that requires too much effort for the moment. I think there are better things to spend time on.

AJenbo commented on source/d1celtilesetframe.cpp:
```diff
@@ -158,3 +160,292 @@ void D1CelTilesetFrame::LoadRightTrapezoid(QByteArray &rawData)
+    int i, x, y;
+    int length = (MICRO_WIDTH * MICRO_HEIGHT) / 2 + MICRO_HEIGHT * (2 + MICRO_HEIGHT) / 4 + MICRO_HEIGHT / 2;
+
+    // memset(pDst, 0, length);
```
I'm not a big fan of the left over code with no comments, it's hard to know if it's old debug code, and alternative metode, something that should work but didn't or something for future rework, or a future optimization?

pionere replied:
The game does not care whether the memory is cleared. It might be useful if you want to compare the results of two tilesets and you want to reduce the number of differences.

AJenbo replied:
In that case lets enable this, it's better to always produce clean dependable data, the game can then figure out if it can ignore things to optimize performance (the editor doesn't have to be particularly well optimized)

pionere replied:
currently it is not needed, because the caller does this.

AJenbo commented on source/d1celtilesetframe.cpp:
```diff
@@ -158,3 +160,292 @@ void D1CelTilesetFrame::LoadRightTrapezoid(QByteArray &rawData)
+        pHead = pDst;
+        pDst++;
+    }
+    // if (!hasColor) {
+    //     qDebug() << "Empty transparent frame"; -- TODO: log empty frame?
+    // }
```
Maybe in the future we could have an option in the editor for optimizing tilesets, where one of the things it does is replace fully transparent frames with a null in the min/til data. In that case I think logging this to `qDebug()` is sufficient.
(potentially we could also run such steps automatically prior to saving)

AJenbo commented on source/d1celtilesetframe.cpp:
```diff
@@ -158,3 +160,292 @@ void D1CelTilesetFrame::LoadRightTrapezoid(QByteArray &rawData)
+{
+    D1CelPixel pixel;
+    int i, x, y;
+    int length = MICRO_WIDTH * MICRO_HEIGHT / 2 + MICRO_HEIGHT;
+
+    // memset(pDst, 0, length);
```
This appears unused?

pionere replied:
you mean the length?

AJenbo replied:
kinda both, length is only used with the commentedout code, and the commentedout isn't doing anything either :P

AJenbo commented on source/d1celtilesetframe.cpp:
```diff
@@ -158,3 +160,292 @@ void D1CelTilesetFrame::LoadRightTrapezoid(QByteArray &rawData)
+    // if (!hasColor) {
+    //     qDebug() << "Empty transparent frame"; -- TODO: log empty frame?
+    // }
+    // pHead = (quint8 *)(((qsizetype)pHead + 3) & (~(qsizetype)3));
```
What is this code for?

pionere replied:
To have 4-byte aligned addresses to the micros. Does not really help, because the 4-byte alignment is messed up anyway (e.g left/right triangles).

AJenbo commented on source/d1cel.cpp:
```diff
@@ -138,3 +140,231 @@ bool D1Cel::load(QString filePath, OpenAsParam *params)
+        }
+    }
+
+    // if (numGroups > 1) {
```
Is this because it should not have been a compilation if there is only 1 group?

pionere replied:
Right, similar to the case above (line 357) I'm not exactly sure how the grouping supposed to work.

pionere commented on source/saveasdialog.cpp:
```diff
@@ -20,8 +21,8 @@ std::optional<QFile> SaveAsParam::getValidSaveOutput(QString filePath, QString s
 
-    QFile outFile = QFile(filePath);
-    if (!outFile.open(QIODevice::WriteOnly | QFile::Truncate)) {
+    QFile *outFile = new QFile(filePath);
```
where is this freed ?

AJenbo commented:
Idealy I should have used `unique_ptr` instead of manually freeing the `QFile`, but I'm not proficient enough to do it by heart so saving that for later when we start applying such cleanups in general.

pionere commented:
> Idealy I should have used `unique_ptr` instead of manually freeing the `QFile`, but I'm not proficient enough to do it by heart so saving that for later when we start applying such cleanups in general.

I think the more 'Qt'-way would be to return a QFile in any case and check if it is `isNull()`.

AJenbo commented:
I'm refering to freein it, not checking if it was successful
